cloud/scope/machine_pool.go

```diff
@@ -0,0 +1,510 @@
+		}
+	}
+
+	return placements, nil
+}
+
+// IsResourceCreatedByClusterAPI determines if the instance was created by the cluster using the
+// tags created at instance launch.
+func (s *MachinePoolScope) IsResourceCreatedByClusterAPI(resourceFreeFormTags map[string]string) bool {
+	tagsAddedByClusterAPI := ociutil.BuildClusterTags(string(s.OCICluster.GetOCIResourceIdentifier()))
+	for k, v := range tagsAddedByClusterAPI {
+		if resourceFreeFormTags[k] != v {
+			return false
+		}
+	}
+	return true
+}
+
+// GetFreeFormTags gets the free form tags for the MachinePoolScope cluster and returns them
+func (m *MachinePoolScope) GetFreeFormTags() map[string]string {
+
+	tags := ociutil.BuildClusterTags(m.OCICluster.GetOCIResourceIdentifier())
+	if m.OCICluster.Spec.FreeformTags != nil {
+		for k, v := range m.OCICluster.Spec.FreeformTags {
+			tags[k] = v
+		}
+	}
+
+	return tags
+}
+
+// ReconcileInstanceConfiguration works to try to reconcile the state of the instance configuration for the cluster
+func (m *MachinePoolScope) ReconcileInstanceConfiguration(ctx context.Context) error {
+	var instanceConfiguration *core.InstanceConfiguration
+	instanceConfigurationId := m.GetInstanceConfigurationId()
+
+	if len(instanceConfigurationId) > 0 {
+		req := core.GetInstanceConfigurationRequest{InstanceConfigurationId: common.String(instanceConfigurationId)}
+		instanceConfiguration, err := m.ComputeManagementClient.GetInstanceConfiguration(ctx, req)
+		if err == nil {
+			m.Info("instance configuration found", "InstanceConfigurationId", instanceConfiguration.Id)
+			m.SetInstanceConfigurationIdStatus(instanceConfigurationId)
+			return m.PatchObject(ctx)
+		} else {
+			return errors.Wrap(err, fmt.Sprintf("error getting instance configuration by id %s", instanceConfigurationId))
+		}
+	}
+
+	if instanceConfiguration == nil {
+		m.Info("Create new instance configuration")
+
+		tags := m.GetFreeFormTags()
+
+		cloudInitData, err := m.GetBootstrapData()
+		if err != nil {
+			return err
+		}
+
+		metadata := m.OCIMachinePool.Spec.Metadata
+		if metadata == nil {
+			metadata = make(map[string]string)
+		}
+		metadata["user_data"] = base64.StdEncoding.EncodeToString([]byte(cloudInitData))
+
+		subnetId := m.GetWorkerMachineSubnet()
+		nsgId := m.GetWorkerMachineNSG()
+
+		launchInstanceDetails := core.ComputeInstanceDetails{
+			LaunchDetails: &core.InstanceConfigurationLaunchInstanceDetails{
+				CompartmentId: common.String(m.OCICluster.Spec.CompartmentId),
+				DisplayName:   common.String(m.OCIMachinePool.GetName()),
+				Shape:         common.String(m.OCIMachinePool.Spec.InstanceConfiguration.InstanceDetails.Shape),
+				SourceDetails: core.InstanceConfigurationInstanceSourceViaImageDetails{
+					ImageId: common.String(m.OCIMachinePool.Spec.ImageId),
+				},
+				Metadata: metadata,
+				CreateVnicDetails: &core.InstanceConfigurationCreateVnicDetails{
+					SubnetId: subnetId,
+					// TODO variablize AssignPublicIp in the future
```

Comment:
plesase add it toe he next Pr ticket, I believe the one where we have to set ocids of instances

Reply:
will be resolved in https://github.com/oracle/cluster-api-provider-oci/issues/90